src/core/util/http.cc

```diff
@@ -109,87 +114,255 @@ bool HTTP::Download(
     // Set SSL options
     options
       .openssl_certificate(cert_path.string())
-      .openssl_sni_hostname(uri.m_Host);
+      .openssl_sni_hostname(uri.host());
   }
+  // Create client with options
+  Client client(options);
   try {
-    // Set extra options
-    options.timeout(45);  // Java I2P defined
-    http::client client(options);
-    // Prepare and initiate session
-    http::client::request request(address);
-    request << network::header("User-Agent", "Wget/1.11.4");  // Java I2P defined
-    http::client::response response = client.get(request);
-    // Assign stream our downloaded contents
-    m_Stream.assign(http::body(response));
-  } catch (const std::exception& e) {
-    LogPrint(eLogError, "HTTP: unable to complete download: ", e.what());
+    // Create request
+    Request request(uri.string());  // A fully-qualified, completed URI
+    // Add required Java I2P defined user-agent
+    request << boost::network::header("User-Agent", "Wget/1.11.4");
+    // Are we requesting the same file?
+    if (uri.path() == GetPreviousPath()) {
+      // Add ETag and Last-Modified headers if previously set
+      if (!GetPreviousETag().empty())
+        request << boost::network::header("If-None-Match", GetPreviousETag());
+      if (!GetPreviousLastModified().empty())
+        request << boost::network::header("If-Modified-Since", GetPreviousLastModified());
+    } else {
+      // Set path to test against for future download (if this is a single instance)
+      SetPath(uri.path());
+    }
+    // Create response object, send request and receive response
+    Response response = client.get(request);
+    // Test HTTP response status code
+    switch (response.status()) {
+      case 200:  // OK  (also, cached version does not match, so we download)
```

Comment:
cpp-netlib may have this readily available. If not, I'll create constants and then a TODO to create a container if we expand use-case.